logins-sql/src/db.rs

```diff
@@ -0,0 +1,789 @@
+            // existing iOS sqlcipher databases).
+            format!("PRAGMA key = '{}';", escape_string_for_pragma(key))
+        } else {
+            "".to_owned()
+        };
+
+        // `temp_store = 2` is required on Android to force the DB to keep temp
+        // files in memory, since on Android there's no tmp partition. See
+        // https://github.com/mozilla/mentat/issues/505. Ideally we'd only
+        // do this on Android, or allow caller to configure it.
+        let initial_pragmas = format!("
+            {}
+            PRAGMA temp_store = 2;
+        ", encryption_pragmas);
+
+        db.execute_batch(&initial_pragmas)?;
+
+        let max_var_limit = db.limit(limits::Limit::SQLITE_LIMIT_VARIABLE_NUMBER);
+        // This is an i32, so check before casting it. We also disallow 0
+        // because it's not clear how we could ever possibly handle it.
+        // (Realistically, we'll also fail to handle low values, but oh well).
+        assert!(max_var_limit > 0,
+                "SQLITE_LIMIT_VARIABLE_NUMBER must not be 0 or negative!");
+
+        let mut logins = Self {
+            db,
+            max_var_count: max_var_limit as usize
+        };
+        schema::init(&mut logins)?;
+        Ok(logins)
+    }
+
+    pub fn open(path: impl AsRef<Path>, encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open(path)?, encryption_key)?)
+    }
+
+    pub fn open_in_memory(encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open_in_memory()?, encryption_key)?)
+    }
+
+    pub fn execute_all(&self, stmts: &[&str]) -> Result<()> {
+        for sql in stmts {
+            self.execute(sql, &[])?;
+        }
+        Ok(())
+    }
+
+    #[inline]
+    pub fn execute(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, false)?)
+    }
+
+    #[inline]
+    pub fn execute_cached(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, true)?)
+    }
+
+    fn do_exec(&self, sql: &str, params: &[(&str, &ToSql)], cache: bool) -> Result<usize> {
+        let res = if cache {
+            self.db.prepare_cached(sql)
+                   .and_then(|mut s| s.execute_named(params))
+        } else {
+            self.db.execute_named(sql, params)
+        };
+        if let Err(e) = &res {
+            warn!("Error running SQL {}. Statement: {:?}", e, sql);
+        }
+        Ok(res?)
+    }
+
+    pub fn query_one<T: FromSql>(&self, sql: &str) -> Result<T> {
+        let res: T = self.db.query_row(sql, &[], |row| row.get(0))?;
+        Ok(res)
+    }
+
+    // The type returned by prepare/prepare_cached are different, but must live
+    // to the end of the function, so (AFAICT) it's difficult/impossible to
+    // remove the duplication between query_row/query_row_cached.
```

Comment:
Maybe wrapping the statements in different anonymous structs, and having those structs implement a trait that calls `query_named`? But...I feel like that would be more code, and more indirection, than duplicating these lines.